src/main/java/com/iota/iri/service/tipselection/impl/RecursiveWeightCalculator.java

```diff
@@ -0,0 +1,147 @@
+
+/**
+ * Calculates the weight recursively/on the fly instead of building the tree and calculating after
+ */
+public class RecursiveWeightCalculator implements RatingCalculator {
+
+    private final Tangle tangle;
+    private final SnapshotProvider snapshotProvider;
+    
+    private Map<Hash, ArrayDeque<Hash>> txToDirectApprovers = new HashMap<>();
+
+    /**
+     * Constructor for Recursive Weight Calculator
+     * @param tangle Tangle object which acts as a database interface
+     * @param snapshotProvider accesses ledger's snapshots
+     */
+    public RecursiveWeightCalculator(Tangle tangle, SnapshotProvider snapshotProvider) {
+        this.tangle = tangle;
+        this.snapshotProvider = snapshotProvider;
+    }
+
+    @Override
+    public UnIterableMap<HashId, Integer> calculate(Hash entryPoint) throws Exception {
+        UnIterableMap<HashId, Integer> hashWeight = calculateRatingDfs(entryPoint);
+        
+        return hashWeight;
+    }
+    
+    private UnIterableMap<HashId, Integer> calculateRatingDfs(Hash entryPoint) throws Exception {
+        TransactionViewModel tvm = TransactionViewModel.fromHash(tangle, entryPoint);
+        int depth = tvm.snapshotIndex() > 0 
+                ? snapshotProvider.getLatestSnapshot().getIndex() - tvm.snapshotIndex() + 1 
+                : 1;
+
+        // Estimated capacity per depth, assumes 5 minute gap in between milestones, at 3tps
+        UnIterableMap<HashId, Integer> hashWeight = createTxHashToCumulativeWeightMap( 5 * 60 * 3 * depth);
+        
+        Deque<Hash> stack = new ArrayDeque<>();
+        Map<Hash, ArrayDeque<Hash>> txToDirectApprovers = new HashMap<>();
+
+        stack.push(entryPoint);
+        while (CollectionUtils.isNotEmpty(stack)) {
+            Hash txHash = stack.peek();
+            if (!hashWeight.containsKey(txHash)) {
+                Collection<Hash> appHashes = getTxDirectApproversHashes(txHash, txToDirectApprovers, txToDirectApprovers);
+                if (CollectionUtils.isNotEmpty(appHashes)) {
+                    Hash txApp = getAndRemoveApprover(appHashes);
+                    stack.push(txApp);
+                    continue;
+                }
+            } else {
+                stack.pop();
+                continue;
+            }
+            
+            HashSet<HashId> set = new HashSet<>();
+            set.add(txHash);
+            
+            int rating = txHash.equals(entryPoint) ? hashWeight.size() + 1 : getRating(txHash, set);
+            hashWeight.put(txHash, rating);
+        }
+        return hashWeight;
+    }
+
+    private int getRating(Hash hash, Set<HashId> seenHashes) throws Exception {
+        int weight = 1;
+
+        ArrayDeque<Hash> approvers = getTxDirectApproversHashes(hash, txToDirectApprovers, null);
+        for (Hash approver : approvers) {
+            if (!seenHashes.contains(approver)) {
+                seenHashes.add(approver);
+                weight += getRating(approver, seenHashes);
```

Comment:
It is better to have something iterative rather than recursive calls. May prevent `StackOverflows`...
You are simply doing a `dfs`, so you can reuse the code of `calculateRatingDfs`